src/instructlab/config/init.py

```diff
@@ -183,6 +209,70 @@ def init(
     )
 
 
+def hw_auto_detect() -> tuple[str | None, int | None, _train, bool]:
+    # Third Party
+    import torch
+
+    gpus = 0
+    total_vram = 0
+    train = _train()
+    matched_gpu_count = False
+    chosen_vram = None
+    chosen_profile_name = None
+    gpu_name = ""
+    # try nvidia
+    if torch.cuda.is_available():
+        click.echo("Detecting Hardware...")
+        gpus = torch.cuda.device_count()
+        for i in range(gpus):
+            properties = torch.cuda.get_device_properties(i)
+            gpu_name = properties.name
+            total_vram += properties.total_memory  # memory in B
+    vram = int(floor(convert_bytes_to_proper_mag(total_vram)[0]))
+    if vram == 0:
+        return None, None, train, False
+    profiles = PROFILE_MAPPINGS
+    for group_gpu_name, list_of_count_vram_and_config in profiles.items():
```

Comment:
I would recommend going with a lookup approach which should simplify the code further.  Two possible approaches:

Default Fallback Approach

```
def lookup_card(gpu_name, gpu_count, vram):
     card_entry = PROFILE_MAPPINGS.get(gpu_name, UNKOWN_NVIDIA)
     # select_card will examine the tuple/list for a matching gpu count and card vram entry (if variants)
     return select_best_card_config(card_entry, gpu_count, vram)
```

UNKNOWN_NVIDIA would point to a tuple/list that would contain the lowest common denominator on gpu capabilities with a number of entries on GPUs and sizes, this might be an alias to a specific config for this purpose or it could be an alias to an another key

Scan All Fallback

Alternatively could scan the table similar to this code but with the direct lookup and a tighter scan loop:

```
def lookup_card(gpu_name, gpu_count, vram):
     card_entry = PROFILE_MAPPINGS.get(gpu_name, UNKOWN_NVIDIA)
     # select_card will examine the tuple/list for a matching gpu count and card vram entry (if variants)
     if card_entry is not None:
          return select_best_card_config(card_entry, gpu_count, vram)

    return find_closest_gpu_vram_match(gpu_count, vram)
```

Reply:
ok, thanks jason! I will implement one of the above

Reply:
there is one issue with:
```
def lookup_card(gpu_name, gpu_count, vram):
     card_entry = PROFILE_MAPPINGS.get(gpu_name)
     # select_card will examine the tuple/list for a matching gpu count and card vram entry (if variants)
     if card_entry is not None:
          return select_best_card_config(card_entry, gpu_count, vram)

    return find_closest_gpu_vram_match(gpu_count, vram)
    ```
    it is not a sure bet that just because we match the GPU name we find a profile with vRAM and GPU count. So there is a scenario where we fall back on the vram based match

this is fine with me, but its basically the same end result as what I currently have

I am going to push a version with these abstractions soon though because it helps to have sep functions

Reply:
switched to this structure, PTAL!